ring/watcher.go

```diff
@@ -0,0 +1,54 @@
+package ring
+
+import (
+	"fmt"
+	"sync"
+)
+
+type Watcher interface {
+	Register() (<-chan interface{}, error)
+	UnRegister()
+	notify()
+}
+
+type watcher struct {
+	mtx      sync.Mutex
+	notifier chan interface{}
+}
+
+func (s *watcher) Register() (<-chan interface{}, error) {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	if s.notifier != nil {
+		return nil, fmt.Errorf("only one watcher allowed at a time")
+	}
+	s.notifier = make(chan interface{})
+	return s.notifier, nil
+}
+
+func (s *watcher) UnRegister() {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	if s.notifier != nil {
+		close(s.notifier)
+	}
+	s.notifier = nil
+}
+
+// Handles notifying listeners about state change in the Ring
+// Assumes that the callee will check that there is meaningfull change in the states
```

Comment:
I think the correct term would be "the caller", i.e. the caller of the method:
```suggestion
// Assumes that the caller will check that there is meaningful change to the state.
```

Reply:
depends on how you look at it I suppose. I'll change it to reflect the user experience more

Reply:
The callee would be this method, wouldn't it since it's getting called? It's the caller of the method that should ensure there's meaningful state change right?